_drafts/device-logging.md

```diff
@@ -0,0 +1,273 @@
+
+<p align="center">
+  <img width="850" src="{% img_url blog-device-logging/jit-logs.png %}" alt="Just in time log collection" />
+</p>
+
+## Logging practices
+
+### Write logs for the target audience
+
+It’s important to consider the target audience of your logs, especially as the use of your device grow over the years, and more people and more teams are involved in its development.
+
+When you’re only logging for yourself — really, anything goes, as you’ll have all the context to make interpreting the logs easier (though you might still want to give your future self an easier time too). But when your logs might be used by your teammates, or even other teams with varying context and skills, you might want to use some strategies to make sure that the logs are also effective for them.
+
+#### Make logs actionable
+
+During development, we can be guilty of logging in a way that’s technically accurate rather than actionable, because we assume the eventual reader of the log will have as much context and knowledge of the domain as we do, and they’ll be able to diagnose the remedy themselves.
+
+```kotlin
+when (val name = reader.nextName()) {
+ VERSION -> Logger.w("Invalid key: '$name'")
+  else -> Logger.w("Unhandled key: '$name'")
+}
+```
+
+Instead of only adding the technical reason for the failure, I’ve found it helpful to add information about how the error is expected to be resolved, in addition to the domain specific information about what went wrong [^meaningfulmessages].
+
+```kotlin
+when (val name = reader.nextName()) {
+ VERSION -> Logger.w(
+    "Invalid key: '$name', the key should only be written, not read"
+  )
+  else -> Logger.w(
+  "Unhandled key: '$name', is the parser version behind the writer version?"
+ )
+}
+```
+
+Adding additional context can jumpstart the debugging effort, as now the person debugging has an avenue to investigate immediately after seeing the log.
+
+#### Link logs to additional documentation
+
+It doesn't feel right to write a paragraph in the log, but it can often take a few sentences to actually explain what the error being logged is.
+
+```kotlin
+try {
+  ...
+} catch (e: RemoteException) {
+  Logger.w("Unable to connect to Reporter, this happens if Reporter is so old that it does not contain the Service at all.", e)
+}
+```
+
+Instead of logging a large paragraph, we can log a reference link to somewhere else where the paragraph is written. This is similar to how SDKs will also contain a glossary of all the exceptions it may throw, what each error means, and documentation on how to resolve each one. We can apply that same logic with our own logging.
+
+```kotlin
+try {
+  ...
+} catch (e: RemoteException) {
+  // This can happen if Reporter is so old that it does not contain the Service at all
```

Comment:
Super nit: How about removing this comment? A few reasons:
1. I see the use of the reference links not only making logs more readable, but also making the source (comments + code!) more concise. Just shifting the contents of the log message into a comment above defeats that purpose.
2. We are suggesting that a glossary of terms should be created, which will define `REPORTER_SERVICE_NOT_FOUND`.  In that case, I think it is better to have one source of truth and avoid extra comments like this.

Reply:
Yeah I tried to have kind of 3 examples here to show moving from a long log to a comment to additional docs but maybe a bit unnecessary.